File: Grade_2/2-1/DS/hw3/rhkwp3.py
import time

MAX_WORD = 100
MAX_MEAN = 200

class Node:
    def __init__(self):
        self.word    = ""
        self.meaning = ""
        self.intword = []
        self.num     = 0
        self.prev    = None
        self.next    = None

head  = None
tail  = None
total = 0

def chg_intword(src):
    return [ord(c.upper()) for c in src]

def create_node(word, mean):
    node         = Node()
    node.word    = word[:MAX_WORD - 1]
    node.meaning = mean[:MAX_MEAN - 1]
    node.intword = chg_intword(word)
    return node

def insert_sorted(node):
    global head, tail, total

    curr = head
    while curr is not None and node.intword > curr.intword:
        curr = curr.next

    if curr is None:
        if tail is None:
            head = tail = node
        else:
            node.prev  = tail
            tail.next  = node
            tail       = node
    else:
        node.next = curr
        node.prev = curr.prev

        if curr.prev is not None:
            curr.prev.next = node
        else:
            head = node

        curr.prev = node

    total += 1

def assign_numbers():
    idx_n = 1
    curr  = head
    while curr is not None:
        curr.num = idx_n
        idx_n   += 1
        curr     = curr.next

def search(word):
    iw   = chg_intword(word)
    curr = head
    while curr is not None:
        if iw == curr.intword:
            return curr
        if iw < curr.intword:
            return None
        curr = curr.next
    return None


# ── (2) 개선: 알파벳 첫 글자별 인덱스 테이블 ──────────────────────────────────

alpha_index = {}

def build_index():
    curr = head
    while curr is not None:
        key = curr.word[0].lower()
        if key not in alpha_index:
            alpha_index[key] = curr
        curr = curr.next

def search_improved(word):
    iw   = chg_intword(word)
    key  = word[0].lower()
    curr = alpha_index.get(key, head)
    while curr is not None:
        if iw == curr.intword:
            return curr
        if iw < curr.intword:
            return None
        curr = curr.next
    return None


# ── 메인 ───────────────────────────────────────────────────────────────────────

def main():
    fp = open("randdict_utf8.TXT", "r", encoding="utf-8")

    print("사전 로딩 중...")

    for line in fp:
        line = line.rstrip('\n')

        split = line.find(':')
        if split < 0:
            continue

        word = line[:split].rstrip()
        mean = line[split + 1:].lstrip()

        if not word:
            continue

        insert_sorted(create_node(word, mean))

    fp.close()

    assign_numbers()
    print(f"총 {total}개 단어 로딩 완료.\n")
    print("단어를 입력하세요 (종료: quit)")

    while True:
        word = input(">> ")

        if word == "quit" or word == "exit":
            break
        if not word:
            continue

        r = search(word)
        if r:
            print(f"({r.num}) {r.meaning}")
        else:
            print(f"'{word}'를 찾을 수 없습니다.")


def benchmark():
    global head, tail, total, alpha_index

    head = tail = None
    total = 0
    fp = open("randdict_utf8.TXT", "r", encoding="utf-8")
    for line in fp:
        line  = line.rstrip('\n')
        split = line.find(':')
        if split < 0:
            continue
        word = line[:split].rstrip()
        mean = line[split + 1:].lstrip()
        if not word:
            continue
        insert_sorted(create_node(word, mean))
    fp.close()
    assign_numbers()

    import random
    all_words = []
    curr = head
    while curr:
        all_words.append(curr.word)
        curr = curr.next
    sample = random.sample(all_words, 10000)

    t0 = time.time()
    for w in sample:
        search(w)
    t_base = time.time() - t0

    build_index()
    t0 = time.time()
    for w in sample:
        search_improved(w)
    t_impr = time.time() - t0

    print(f"기존 연결리스트 저장 구조에서 랜덤한 1만개 단어 검색 : {t_base:.3f}초")
    print(f"개선된 연결리스트 저장 구조에서 랜덤한 1만개의 단어 검색 : {t_impr:.3f}초")


## 메인
print("1. 기본 사전  2. 성능 비교")
mode = input("선택: ").strip()
if mode == "2":
    print("사전 로딩 중...")
    benchmark()
else:
    main()

File: Grade_2/2-1/DS/hw3/rhkwp3.c
/*
 * rhkwp3.c - 연결 리스트 사전 (이중 연결리스트)
 *
 * 노드 구성:
 *   word     - 사전 단어 (문자열)
 *   meaning  - 단어의 뜻 (문자열)
 *   intword  - 단어를 대문자 ASCII로 변환한 char 배열 (삽입 위치 비교용)
 *   num      - 알파벳 순 몇 번째 노드인지 (전체 로딩 후 부여)
 *   prev     - 이전 노드 포인터
 *   next     - 다음 노드 포인터
 *
 * 연결리스트 선택 이유:
 *   이중 연결리스트를 선택. 단어 삽입 시 이전/다음 노드에 모두 접근해야
 *   정렬된 위치에 노드를 연결할 수 있기 때문.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#define MAX_WORD  100
#define MAX_MEAN  200

typedef struct Node {
    char  word[MAX_WORD];
    char  meaning[MAX_MEAN];
    char  intword[MAX_WORD];  /* 대문자 ASCII 변환값 배열 (char형) */
    int   num;
    struct Node *prev;
    struct Node *next;
} Node;

static Node *head  = NULL;
static Node *tail  = NULL;
static int   total = 0;

/* 단어를 대문자 ASCII 배열로 변환 */
static void chg_intword(char *src, char *dst) {
    int i;
    for (i = 0; src[i] != '\0' && i < MAX_WORD - 1; i++)
        dst[i] = (char)toupper((unsigned char)src[i]);
    dst[i] = '\0';
}

static Node *create_node(char *word, char *mean) {
    Node *node = (Node *)malloc(sizeof(Node));

    if (node == NULL) { 
        printf("메모리 할당 실패"); 
        return 0; 
    }

    strncpy(node->word, word, MAX_WORD - 1); node->word[MAX_WORD - 1] = '\0';
    strncpy(node->meaning, mean, MAX_MEAN - 1); node->meaning[MAX_MEAN - 1] = '\0';
    chg_intword(word, node->intword);

    node->num  = 0;
    node->prev = node->next = NULL;
    return node;
}

/* intword를 기준으로 알파벳 순 정렬 삽입 */
static void insert_sorted(Node *node) {
    Node *curr = head;

    /* 삽입 위치 탐색: node->intword 보다 큰 첫 번째 노드 앞에 삽입 */
    while (curr && strcmp(node->intword, curr->intword) > 0)
        curr = curr->next;

    if (curr == NULL) {
        /* 리스트가 비어있거나 맨 뒤에 삽입 */
        if (tail == NULL) {
            head = tail = node;
        } 
        else {
            node->prev  = tail;
            tail->next  = node;
            tail = node;
        }
    }
    else {
        /* curr 바로 앞에 삽입 */
        node->next  = curr;
        node->prev  = curr->prev;

        if (curr->prev) 
            curr->prev->next = node;
        else 
            head = node;

        curr->prev  = node;
    }
    total++;
}

/* 전체 로딩 후 head부터 순회하며 번호 부여 */
static void assign_numbers(void) {
    int  idx_n = 1;
    Node *curr = head;
    while (curr) {
        curr->num = idx_n++;
        curr = curr->next;
    }
}

/* intword 기준 순차 탐색 (정렬되어 있으므로 초과 시 조기 종료) */
static Node *search(char *word) {
    char iw[MAX_WORD];
    chg_intword(word, iw);

    Node *curr = head;
    while (curr) {
        int cmp = strcmp(iw, curr->intword);
        if (cmp == 0) 
            return curr;
        if (cmp <  0)
            return NULL;
        curr = curr->next;
    }
    return NULL;
}

static void free_list(void) {
    Node *curr = head;
    while (curr) {
        Node *next = curr->next;
        free(curr);
        curr = next;
    }
    head = tail = NULL;
}

int main(void) {
    FILE *fp = fopen("randdict_utf8.TXT", "r");

    if (fp == NULL) {
        fprintf(stderr, "파일 열기 실패\n");
        return 1;
    }

    char line[MAX_WORD + MAX_MEAN];
    printf("사전 로딩 중...\n");

    while (fgets(line, sizeof(line), fp)) {
        /* 줄 끝 개행 제거 */
        int len = (int)strlen(line);
        while (len > 0 && line[len-1] == '\n'){
            len -= 1;
            line[len] = '\0';
        }

        /* ':' 기준으로 단어/뜻 분리 (첫 번째 ':' 사용) */
        char *split = strchr(line, ':');
        *split = '\0';

        char *word = line;
        char *mean = split + 1; 

        /* 단어 뒤쪽 공백 제거 */
        int wlen = (int)strlen(word);
        while (wlen > 0 && word[wlen-1] == ' ')
            wlen -= 1;
            word[wlen] = '\0';

        /* 뜻 앞쪽 공백 제거 */
        while (*mean == ' ')
            mean += 1;

        insert_sorted(create_node(word, mean));
    }
    fclose(fp);

    assign_numbers();
    printf("총 %d개 단어 로딩 완료.\n\n", total);
    printf("단어를 입력하세요 (종료: quit)\n");

    char input[MAX_WORD];
    while (1) {
        printf(">> ");
        if (fgets(input, sizeof(input), stdin) == NULL)
            break;
        int ilen = (int)strlen(input);

        while (ilen > 0 && (input[ilen-1] == '\n' || input[ilen-1] == '\r'))
            input[--ilen] = '\0';

        if (!strcmp(input, "quit") || !strcmp(input, "exit")) 
            break;
        if (!*input) 
            continue;

        Node *r = search(input);

        if (r) 
            printf("(%d) %s\n", r->num, r->meaning);
        else
            printf("'%s'를 찾을 수 없습니다.\n", input);
    }

    free_list();
    return 0;
}
